src/cmd/flux-job.c

```diff
@@ -919,6 +922,87 @@ void attach_signal_cb (flux_reactor_t *r, flux_watcher_t *w,
     }
 }
 
+static void attach_send_shell_completion (flux_future_t *f, void *arg)
+{
+    /* failng to write stdin to service is a fatal error */
+    if (flux_future_get (f, NULL) < 0) {
+        /* stdin may not be accepted for multiple reasons
+         * - job has completed
+         * - user requested stdin via file
+         * - stdin stream already closed due to prior pipe in
+         */
+        if (errno == ENOSYS)
+            log_msg_exit ("stdin not accepted by job");
+        log_err_exit ("attach_send_shell");
+    }
+    flux_future_destroy (f);
+}
+
+static int attach_send_shell (struct attach_ctx *ctx,
+                              void *buf,
+                              int len,
+                              bool eof)
+{
+    json_t *context = NULL;
+    char topic[1024];
+    flux_future_t *f = NULL;
+    int saved_errno;
+    int rc = -1;
+
+    snprintf (topic, sizeof (topic), "shell-%ju.stdin", (uintmax_t)ctx->id);
+    if (!(context = ioencode ("stdin", "all", buf, len, eof)))
+        goto error;
+    if (!(f = flux_rpc_pack (ctx->h, topic, ctx->leader_rank, 0, "O", context)))
+        goto error;
+    if (flux_future_then (f, -1, attach_send_shell_completion, ctx) < 0)
+        goto error;
+    /* f memory now in hands of attach_send_shell_completion() */
+    f = NULL;
+    rc = 0;
+ error:
+    saved_errno = errno;
+    json_decref (context);
+    flux_future_destroy (f);
+    errno = saved_errno;
+    return rc;
+}
+
+/* Handle std input from user */
+void attach_stdin_cb (flux_reactor_t *r, flux_watcher_t *w,
+                      int revents, void *arg)
+{
+    struct attach_ctx *ctx = arg;
+
+    if (revents & FLUX_POLLIN) {
+        ssize_t n;
+        long ps = sysconf (_SC_PAGESIZE);
+        char buf[ps];
+
+        assert (ps > 0);
+
+        if ((n = read (STDIN_FILENO, buf, ps)) < 0)
+            log_err_exit ("read on stdin");
+
+        if (n > 0) {
+            if (attach_send_shell (ctx, buf, n, false) < 0)
+                log_err_exit ("attach_send_shell");
+            ctx->total_stdin += n;
+        }
+        else {
+            /* The most common scenario is user has no stdin, no need
+             * to send EOF to shell */
+            if (ctx->total_stdin) {
```

Comment:
ahh yes, my mistake was assuming "flux job attach $id" (i.e. no stdin) would lead to an EOF on STDIN, but that's not the case.

Reply:
oh wait, I was initially confused.  I added this predominantly to handle the case of someone not sending stdin.  So for example:

flux job attach $id > 1.out &
flux job attach $id < myin > 2.out &

I wouldn't want the first attach to write EOF, leading to problems for the second.

Is this just a use case I shouldn't consider?

Reply:
Nothing I can see is closing `stdin` in the first command so I don't see why that command would write any EOFs.

Reply:
yeah, something is up in the tests, I'm not sure why I'm seeing an EOF there.

Reply:
I added a stupid printf to see if I was seeing EOF on stdin.

```
        else {
            fprintf (stderr, "%s:%d\n", __FUNCTION__, __LINE__);
            /* The most common scenario is user has no stdin, no need
             * to send EOF to shell */
            if (ctx->total_stdin) {
                if (attach_send_shell (ctx, NULL, 0, true) < 0)
                    log_err_exit ("attach_send_shell");
            }
```

When I run 

```
flux mini run -n1 echo foo
```

in a local flux instance, I don't see the printf (i.e. stdin isn't closed).

But when I run this unit test under sharness

```
test_expect_success 'flux-shell: run 1-task no pipe in stdin' '
        flux mini run -n1 echo foo
'
```

I do see the printf indicating EOF.  I'm wondering if there's some stdin subtlety with sharness, b/c I know its running the test under a fork and under some environment it sets up.

Reply:
Hmm, wondering if

```
test_eval_() {
    # This is a separate function because some tests use
    # "return" to end a test_expect_success block early.
    case ",$test_prereq," in
    *,INTERACTIVE,*)
        eval "$*"
        ;;
    *)
        eval </dev/null >&3 2>&4 "$*"
        ;;
    esac
}
```

in sharness is the culprit.

Reply:
Yeah, that could be it. We may have to do something clever if you want to test that scenario. Perhaps you could use `mkfifo` to simulate a terminal with nothing typed on it...

Reply:
`mkfifo` seems like a good idea.  I was originally thinking we'd need a "no stdin" option like with `flux-exec`.

Reply:
Doesn't `-n` in `flux-exec` redirect `stdin` from `/dev/null` though?

Reply:
> Doesn't -n in flux-exec redirect stdin from /dev/null though?

I actually wasn't sure.  It ends up doing something we don't want for attach.

```
    if (optparse_getopt (opts, "noinput", NULL) > 0) {
        flux_subprocess_t *p;
        p = zlist_first (subprocesses);
        while (p) {
            if (flux_subprocess_close (p, "stdin") < 0)
                log_err_exit ("flux_subprocess_close");
            p = zlist_next (subprocesses);
        }
    }
```

so minimally we can't do that idea.

Reply:
Tried `run_pty()` above and it appears to work.  But the gotch is ...

```
>diff trash-directory.t2607-job-shell-input/pipe4A.out trash-directory.t2607-job-shell-input/input_stdin_file | cat -t
1,2c1,2
< foo^M
< doh^M
---
> foo
> doh
```

:-(

Giving `mkfifo` a shot and I think I got it working.  It took me awhile to realize that some file descriptors numbers I was using were already used by sharness, ugh ...

Reply:
oh crud, I just realized another problem.  While I can do some `mkfifo` tricks in my specific tests, just doing a `flux mini run -n1 echo foo` in sharness would be a problem everywhere in our sharness tests.

Reply:
In addition to our offline discussion, I went back and checked what happens with `flux wreck attach` just for another data point. In `flux wreck attach` stdin watcher is only started if the job is not in the `complete` state. There is still a race if a job transitions to `complete`  in the meantime, however, this isn't a problem in the wreck prototype since input is written directly to the kvs, the data is ignored (which is probably bad depending on your perspective)

For running jobs in sharness, probably stdin was always immediately closed unless otherwise redirected. This didn't affect any test jobs because none of the tests read from stdin anyway (unless specifically testing stdin broadcast/redirection)

Reply:
Was thinking about this on the way home.  Instead of only giving the user a warning if stdin cannot be piped to the job:

A) if the user attempts to write actual data to stdin, and it can't be piped to the job, that should be an error.

B) if the user only closes stdin, and the EOF can't be piped to the job, that should be a warning

I can see the latter case happening often, but the former case feels like an actual error scenario.